File: plugins/voice-fx/telephone/telephone.js
// Telephone — band-passes the signal into the 300 Hz–3.4 kHz range
// (POTS handset bandwidth) and adds a touch of compression-style
// distortion to mimic GSM codec artefacts. Drop on a vocal for the
// classic "caller on the other end of the line" effect.

function TelephoneProcessor() {
  // Two cascaded one-pole filters: HPF then LPF = band-pass.
  this.hpPrev = 0;
  this.hpOut = 0;
  this.lpOut = 0;
}
TelephoneProcessor.parameterDescriptors = [
  { name: 'lowCut',  label: 'Low Hz',  defaultValue: 300,  minValue: 100, maxValue: 1000 },
  { name: 'highCut', label: 'High Hz', defaultValue: 3400, minValue: 1500,maxValue: 8000 },
  { name: 'crunch',  label: 'Crunch',  defaultValue: 0.3,  minValue: 0,   maxValue: 1 },
  { name: 'mix',     label: 'Mix',     defaultValue: 1.0,  minValue: 0,   maxValue: 1 }
];
TelephoneProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0];
  var output = outputs[0][0];
  if (!input || !output) return true;
  var lowCut = parameters.lowCut[0];
  var highCut = Math.max(lowCut + 100, parameters.highCut[0]);
  var crunch = parameters.crunch[0];
  var mix = parameters.mix[0];

  // One-pole coefficients (RC form): smaller alpha → tighter cutoff.
  var dt = 1 / sampleRate;
  var hpRc = 1 / (2 * Math.PI * lowCut);
  var hpA = hpRc / (hpRc + dt);
  var lpRc = 1 / (2 * Math.PI * highCut);
  var lpA = dt / (lpRc + dt);

  for (var i = 0; i < input.length; i++) {
    var x = input[i];
    // HPF
    this.hpOut = hpA * (this.hpOut + x - this.hpPrev);
    this.hpPrev = x;
    // LPF on the HPF output
    this.lpOut = this.lpOut + lpA * (this.hpOut - this.lpOut);
    // Soft saturation for that codec crunch
    var k = 1 + crunch * 12;
    var wet = Math.tanh(this.lpOut * k) / Math.tanh(k);
    output[i] = x * (1 - mix) + wet * mix;
  }
  return true;
};

registerProcessor('telephone', TelephoneProcessor);

File: plugins/voice-fx/telephone/Telephone.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Telephone (native port) — POTS-handset bandpass (~300 Hz – 3.4 kHz)
// plus a soft codec crunch. Classic "caller on the other end" effect.
public final class Telephone implements VocalMonitorNativePlugin {
    private float hpPrev = 0f, hpOut = 0f, lpOut = 0f;
    private int sampleRate = 44100;
    private float lowCut = 300f, highCut = 3400f, crunch = 0.3f, mix = 1f;

    @Override
    public void init(int sr) { this.sampleRate = sr; hpPrev = hpOut = lpOut = 0f; }

    @Override public String[] parameterNames() { return new String[] { "lowCut", "highCut", "crunch", "mix" }; }
    @Override public float parameterMin(String n) {
        switch (n) {
            case "lowCut":  return 100f;
            case "highCut": return 1500f;
            default:        return 0f;
        }
    }
    @Override public float parameterMax(String n) {
        switch (n) {
            case "lowCut":  return 1000f;
            case "highCut": return 8000f;
            default:        return 1f;
        }
    }
    @Override public float parameterDefault(String n) {
        switch (n) {
            case "lowCut":  return 300f;
            case "highCut": return 3400f;
            case "crunch":  return 0.3f;
            case "mix":     return 1f;
            default:        return 0f;
        }
    }
    @Override public String parameterLabel(String n) {
        switch (n) {
            case "lowCut":  return "Low Hz";
            case "highCut": return "High Hz";
            case "crunch":  return "Crunch";
            case "mix":     return "Mix";
            default:        return n;
        }
    }
    @Override public void setParameter(String n, float v) {
        switch (n) {
            case "lowCut":  lowCut = v; break;
            case "highCut": highCut = v; break;
            case "crunch":  crunch = v; break;
            case "mix":     mix = v; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        final float hc = Math.max(lowCut + 100f, highCut);
        final float dt = 1f / sampleRate;
        final float hpRc = (float) (1.0 / (2.0 * Math.PI * lowCut));
        final float hpA = hpRc / (hpRc + dt);
        final float lpRc = (float) (1.0 / (2.0 * Math.PI * hc));
        final float lpA = dt / (lpRc + dt);
        final float k = 1f + crunch * 12f;
        final float kNorm = 1f / (float) Math.tanh(k);
        final float mixLocal = mix;
        final float dry = 1f - mixLocal;
        float hpP = hpPrev, hpO = hpOut, lpO = lpOut;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            float x = input[i];
            hpO = hpA * (hpO + x - hpP);
            hpP = x;
            lpO = lpO + lpA * (hpO - lpO);
            float wet = (float) Math.tanh(lpO * k) * kNorm;
            output[i] = x * dry + wet * mixLocal;
        }
        hpPrev = hpP; hpOut = hpO; lpOut = lpO;
    }
}
